From: Joan Torres Lopez <joantolo@redhat.com>
Date: Thu, 13 Aug 2026 13:18:41 +0200
Subject: gdm/authServicesLegacy: Fix credentialManagers authentication

There were a few guards to ensure authentication only works for
the selectedMechanism, however for credentialManagers, the selectedMechanism
isn't used. That was stopping authenticating through the credentialManager flow.

To fix that:
1. When credentialManager authentication is started
   (_hasAnyCredentialManagerToken()), only allow to start the valid
   credentialManager service (the one that has a token).
2. On _handleOnSecretInfoQuery(), check if there's a valid
   credentialManager before the selectedMechanism guard.

Part-of: <https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/4338>

Origin: https://gitlab.gnome.org/GNOME/gnome-shell/-/commit/920509a
---
 js/gdm/authServicesLegacy.js | 39 +++++++++++++++++++++++++--------------
 1 file changed, 25 insertions(+), 14 deletions(-)

diff --git a/js/gdm/authServicesLegacy.js b/js/gdm/authServicesLegacy.js
index 77f9cec..12adc63 100644
--- a/js/gdm/authServicesLegacy.js
+++ b/js/gdm/authServicesLegacy.js
@@ -253,17 +253,12 @@ export class AuthServicesLegacy extends AuthServices {
     }
 
     _handleOnSecretInfoQuery(serviceName, secretQuestion) {
-        if (serviceName !== this._selectedMechanism?.serviceName)
+        // Try to auto-fill with credential manager token first
+        if (this._tryCredentialManagerAutoFill(serviceName))
             return;
 
-        let token = null;
-        if (this._credentialManagers[serviceName])
-            token = this._credentialManagers[serviceName].token;
-
-        if (token) {
-            this._handleAnswerQuery(serviceName, token);
+        if (serviceName !== this._selectedMechanism?.serviceName)
             return;
-        }
 
         this.emit('ask-question', {
             serviceName,
@@ -273,6 +268,20 @@ export class AuthServicesLegacy extends AuthServices {
         });
     }
 
+    _tryCredentialManagerAutoFill(serviceName) {
+        const credentialManager = this._credentialManagers[serviceName];
+        if (!credentialManager)
+            return false;
+
+        const token = credentialManager.token;
+        if (!token)
+            return false;
+
+        this._userVerifier.call_answer_query(
+            serviceName, token, this._cancellable).catch(logErrorUnlessCancelled);
+        return true;
+    }
+
     _handleOnConversationStopped(serviceName) {
         if (serviceName !== this._selectedMechanism?.serviceName &&
             serviceName !== FINGERPRINT_SERVICE_NAME)
@@ -328,13 +337,8 @@ export class AuthServicesLegacy extends AuthServices {
     }
 
     _handleOnVerificationComplete(serviceName) {
-        if (serviceName !== this._selectedMechanism?.serviceName)
-            return;
-
-        if (this._credentialManagers[serviceName]) {
+        if (this._credentialManagers[serviceName])
             this._credentialManagers[serviceName].token = null;
-            this._selectedMechanism = null;
-        }
     }
 
     _handleOnChoiceListQuery(serviceName, promptMessage, list) {
@@ -363,12 +367,19 @@ export class AuthServicesLegacy extends AuthServices {
     }
 
     _handleCanStartService(serviceName) {
+        if (this._hasAnyCredentialManagerToken())
+            return this._credentialManagers[serviceName]?.token !== null;
+
         return serviceName === this._selectedMechanism?.serviceName ||
             (serviceName === FINGERPRINT_SERVICE_NAME &&
             this._enabledMechanisms.some(m => m.serviceName === serviceName) &&
             this._userName);
     }
 
+    _hasAnyCredentialManagerToken() {
+        return Object.values(this._credentialManagers).some(cm => cm.token !== null);
+    }
+
     addCredentialManager(serviceName, credentialManager) {
         if (this._credentialManagers[serviceName])
             return;
